documentation/jetty-documentation/src/main/asciidoc/programming-guide/arch-threads.adoc

```diff
@@ -0,0 +1,200 @@
+
+[plantuml]
+----
+skinparam backgroundColor transparent
+
+compact concise "Thread 1" as WT1
+compact concise "Thread 2" as WT2
+compact concise "Thread 3" as WT3
+compact concise "Thread 4" as WT4
+hide time-axis
+
+@WT1
+0 is T1 #lightgreen
+1 is "Run T1" #dodgerblue
+5 is {hidden}
+
+@WT2
+1 is T2 #lightgreen
+2 is "Run T2" #dodgerblue
+4 is T5 #lightgreen
+5 is "Run T5" #dodgerblue
+10 is {hidden}
+
+@WT3
+2 is T3 #lightgreen
+3 is "Run T3" #dodgerblue
+6 is {hidden}
+
+@WT4
+3 is T4 #lightgreen
+4 is "Run T4" #dodgerblue
+8 is {hidden}
+
+----
+
+This mode may operate like xref:pg-arch-threads-execution-strategy-pc[`Produce-Consume`] when the take over production task run, for example, by thread `Thread 3` takes time to be executed (for example, in a busy server): thread `Thread 2` will produce one task and run it, then produce another task and run it, etc.
+By the time thread `Thread 3` takes over task production from `Thread 2`, all the work might already be done.
+
+This mode may also operate similarly to xref:pg-arch-threads-execution-strategy-pec[`Produce-Execute-Consume`] when the take over production task always finds a free CPU core immediately (for example, in a mostly idle server): thread `Thread 1` will produce a task, yield production to `Thread 2` while `Thread 1` is running the task; `Thread 2` will produce a task, yield production to `Thread 3` while `Thread 2` is running the task, etc.
+
+Differently from `Produce-Execute-Consume`, here production happens on different threads, but the advantage is that the task is run by the same thread that produced it (which is CPU core cache efficient).
+
+[[pg-arch-threads-execution-strategy-adaptive]]
+===== Adaptive Execution Strategy
+The modes of task consumption discussed above are captured by the `org.eclipse.jetty.util.thread.ExecutionStrategy` interface, with an additional implementation that also takes into account the behavior of the task when the task implements `Invocable`.
+
+For example, a task that declares itself as non-blocking could be consumed using the `Produce-Consume` mode, since there is no risk to stop production because the task will not block.
+
+Conversely, a task that declares itself as blocking will stop production, and therefore must be consumed using either the `Produce-Execute-Consume` mode or the `Execute-Produce-Consume` mode.
+Deciding between these two modes depends on whether there is a free thread immediately available to take over production, and this is captured by the `org.eclipse.jetty.util.thread.TryExecutor` interface.
+
+An implementation of `TryExecutor` could be asked whether it can immediately allocate a free thread to run a task, as opposed to a normal `Executor`, that can only queue the task in the hope that there will be a thread in the near future that could run the task.
+
```

Comment:
```suggestion

If a task declares that it can run in either blocking or non blocking mode, then the strategy may choose to use Produce-Consume with the additional invocation flag to force non-blocking mode; or it may choose one of the other strategies used for blocking tasks.
```

Reply:
I left this out on purpose to avoid to go too deep into the implementation details.
For example, I do not mention the EITHER invocation type, the PRODUCE_INVOKE_CONSUME mode, etc.
For now I'd leave it out.

Reply:
I think it is important to add at least a small mention like this. Else it appears that the documentation does not describe the actual implementation. It doesn't need a deep dive, just a mention.

Reply:
Your suggestion is just too obscure for those that don't know the implementation.
99.999% of usages will be either blocking or non-blocking, "either" is really for super-special cases and if a developer is there, reading the code/javadocs is better than reading the documentation.

Reply:
I still think that EITHER and PIC needs to be mentioned somewhere in the doco, otherwise it looks like it is for a different version of the code.